### PR TITLE
Group tensorboard metrics

### PR DESCRIPTION
**Pull request by VictorSanh**
As discussed in issue #38:

- I grouped the metrics into a few groups
- Added speed metrics (under `iteration-time`)
- Added the mapping iteration (or step) vs samples

A preview of the tensorboard on a dummy training can be found [here](https://tensorboard.dev/experiment/54F4T0NVREO5RbLhbIMPcQ).

### CODE REVIEW

File: megatron/global_vars.py [hunk 1]
```diff
@@ -244,7 +244,7 @@ def write(self, names, writer, iteration, normalizer=1.0, reset=False):
         assert normalizer > 0.0
         for name in names:
             value = self.timers[name].elapsed(reset=reset) / normalizer
-            writer.add_scalar(name + '-time', value, iteration)
+            writer.add_scalar('time/' + name + '-time', value, iteration)
```
Comment (stas00) [inline megatron/global_vars.py:247]:
```suggestion
            writer.add_scalar(f'time/{name}-time', value, iteration)
```

if we are adding `(s)` in other time logs, should there be one here? Or nowhere and we know it's always secs?
The other place you added it is in `iteration-time/iteration-time (s)`

Also suggesting a consistent use of f-strings.

File: megatron/training.py [hunk 2]
```diff
@@ -816,16 +820,16 @@ def evaluate_and_print_results(prefix, forward_step_func,
         ppl = math.exp(min(20, total_loss_dict[key].item()))
         string += '{} PPL: {:.6E} | '.format(key, ppl)
         if writer and is_last_rank():
-            writer.add_scalar('{} validation'.format(key),
+            writer.add_scalar('lm-loss-validation/{} validation'.format(key),
```
Comment (stas00) [inline megatron/training.py:823]:
```suggestion
            writer.add_scalar(f'lm-loss-validation/{key} validation',
```

File: megatron/training.py [hunk 3]
```diff
@@ -816,16 +820,16 @@ def evaluate_and_print_results(prefix, forward_step_func,
...
                               total_loss_dict[key].item(),
                               iteration)
-            writer.add_scalar('{} validation vs samples'.format(key),
+            writer.add_scalar('lm-loss-validation/{} validation vs samples'.format(key),
```
Comment (stas00) [inline megatron/training.py:826]:
```suggestion
            writer.add_scalar(f'lm-loss-validation/{key} validation vs samples',
```

File: megatron/training.py [hunk 4]
```diff
@@ -816,16 +820,16 @@ def evaluate_and_print_results(prefix, forward_step_func,
...
                               total_loss_dict[key].item(),
                               args.consumed_train_samples)
             if args.log_validation_ppl_to_tensorboard:
-                writer.add_scalar('{} validation ppl'.format(key), ppl,
+                writer.add_scalar('lm-loss-validation/{} validation ppl'.format(key), ppl,
```
Comment (stas00) [inline megatron/training.py:830]:
```suggestion
                writer.add_scalar(f'lm-loss-validation/{key} validation ppl', ppl,
```

File: megatron/training.py [hunk 5]
```diff
@@ -816,16 +820,16 @@ def evaluate_and_print_results(prefix, forward_step_func,
...
                                   iteration)
-                writer.add_scalar('{} validation ppl vs samples'.format(key),
+                writer.add_scalar('lm-loss-validation/{} validation ppl vs samples'.format(key),
```
Comment (stas00) [inline megatron/training.py:832]:
```suggestion
                writer.add_scalar(f'lm-loss-validation/{key} validation ppl vs samples',
```

File: megatron/training.py [hunk 6]
```diff
@@ -582,10 +584,12 @@ def add_to_logging(name):
     if iteration % args.log_interval == 0:
         elapsed_time = timers('interval-time').elapsed()
         elapsed_time_per_iteration = elapsed_time / total_iterations
-        if writer and torch.distributed.get_rank() == 0:
+        if writer and is_last_rank(): # Changed `torch.distributed.get_rank() == 0` to `is_last_rank()` as a writing condition
```
Comment (stas00) [inline megatron/training.py:587]: IMHO, it's more important to document why this is done. We have `git diff` to show what was here before.

**Review by stas00 — APPROVED** [review]
Thank you for implementing this, @VictorSanh 

I proposed to use f-strings consistently.

And also a few small questions.

File: megatron/global_vars.py (hunk 1 above)
Comment (VictorSanh) [reply megatron/global_vars.py:247]: yes, as far as i understand, all time-related metrics are reported in seconds, so removing these `(s)`

**Comment by stas00** [comment]
btw, if you go to "Files Changed" tab, you can merge multiple suggestions in one click, so you then need to click half as much ;)

File: megatron/training.py (hunk 6 above)
Comment (VictorSanh) [reply megatron/training.py:587]: i dived deeper, i think it's an error. opened an issue on the megatron-lm side -> https://github.com/NVIDIA/Megatron-LM/issues/129

**Comment by VictorSanh** [comment]
> btw, if you go to "Files Changed" tab, you can merge multiple suggestions in one click, so you then need to click half as much ;)

ooooh, wasn't aware of that, learned something!

**Comment by VictorSanh** [comment]
merging now, thank you @stas00 for the review!

**Comment by stas00** [comment]
BTW, that's not what I meant. Perhaps let's give a bit of time to have a chance to review changes in the future before merging.

`# bug fix in xyz` comments are great inside tests since it's then self-explanatory of what the test is testing.

Such comments in the code have no added value and only make the code harder to read. All of software is ridden with bugs, the normal (as compared to tests) code comments are useful when they explain **why something is done** when it's not obvious from reading the code. 

So when I made that comment I was proposing to add a comment why it was `is_last_rank()`, if it wasn't obvious from the code, which it is not, which you explained in your Issue report to Megatron. This is what we want in the code. Currently the way it's committed the user has to go and research why. Whereas the only comment needed there was:

> # only the last rank process has a non-None _GLOBAL_TENSORBOARD_WRITER.

Hope my logic makes sense.

**Comment by VictorSanh** [comment]
Thanks for flagging that @stas00, always good to write better comments!
I fixed that with your suggestion (5e3963db4a5db36e1ebdb152e2760382bdc7ef04).

I initially merged since you approved the changes. Usually, I use "Request changes" when things are bugging me and "Approved" as a final signal that the PR can be merged when things have been fixed. Does this make any sense to you?

**Comment by stas00** [comment]
Thank you for fixing the comment, Victor

This is the workflow pattern we use at `transformers`, a general Approval is made when things are more or less correct and then the nits can take some time to sort out. This always works when the PR writer is not a committer too. When it's between us the committers we usually check in again when we feel it's ready to be merged.

We can try a different pattern here, and not to Approve until it's ready to be merged. I think this is a better communication style. Over at `transformers` an early Approval I think helps for the PR writer to know they are "out of the woods" and finish line is near.

**Pull request merged**